HPCC-16883 Add time zone support to Std.Date

dcamper opened the pull request:
Signed-off-by: Dan S. Camper <dan.camper@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->
Additional tests have been added to ecllibrary/teststd/TestDate.ecl

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

dcamper committed:
HPCC-16883 Add time zone support to Std.Date

Signed-off-by: Dan S. Camper <dan.camper@lexisnexisrisk.com>

richardkchapman commented:
@JamesDeFabia @ghalliday Please review

JamesDeFabia reviewed (commented):
Looks good from my POV. Excellent javadoc comments. @ghalliday should have the last word

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: e7d12526372e132186c2b077c7df617d84a4cdbe
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |82 |82 |0 |0 |0 |
|wutoolTest(Dali) |19 |17 |2 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |




Wutool error(s):
Dali:1) test: WuTool::testList (F) line: 1656 /mnt/disk1/home/vamosax/smoketest/PR-10505/HPCC-Platform/tools/wutool/wutool.cpp

Dali:equality assertion failed

Dali:2) test: WuTool::testSortByThorTime (F) line: 1930 /mnt/disk1/home/vamosax/smoketest/PR-10505/HPCC-Platform/tools/wutool/wutool.cpp

Dali:assertion failed


Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |650 |650 |0 |
|test (roxie) |768 |768 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   24 sec (00:00:24) |  179 sec (00:02:59) |   55 sec (00:00:55) |    1 sec (00:00:01) |   43 sec (00:00:43) | 1385 sec (00:23:05) |   57 sec (00:00:57) | 1744 sec (00:29:04) |

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: e7d12526372e132186c2b077c7df617d84a4cdbe
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |82 |82 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |740 |1 |
|test (thor) |650 |649 |1 |
|test (roxie) |768 |767 |1 |


[Errors:](https://gist.github.com/c44a3a8f6c2f256033ab5ca77dee1eb3)
- test
  - hthor
    - 667\. [teststdlibrary](https://gist.github.com/HPCCSmoketest/c44a3a8f6c2f256033ab5ca77dee1eb3#file-teststdlibrary-w20180214-130429-5-log), 
  - thor
    - 595\. [teststdlibrary](https://gist.github.com/HPCCSmoketest/c44a3a8f6c2f256033ab5ca77dee1eb3#file-teststdlibrary-w20180214-130915-2-log), 
  - roxie
    - 699\. [teststdlibrary](https://gist.github.com/HPCCSmoketest/c44a3a8f6c2f256033ab5ca77dee1eb3#file-teststdlibrary-w20180214-131529-log), 

HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    5 sec (00:00:05) |  177 sec (00:02:57) |   55 sec (00:00:55) |    6 sec (00:00:06) |   37 sec (00:00:37) | 1093 sec (00:18:13) |   32 sec (00:00:32) | 1405 sec (00:23:25) |

AttilaVamos commented:
@richardkchapman Are those teststdlibrary failures 
```
(new warnings:
+<Warning><Code>4531</Code><Filename>Date.ecl</Filename><Line>1893</Line><Source>eclcc</Source><Message>JOIN condition folded to constant, converting to an ALL join</Message></Warning>) 
```
come from your change?

dcamper commented:
@richardkchapman I believe so, yes.  The new vTZInputSample attribute needs a NOFOLD to prevent constant folding.  I'll make that change.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 059d35b562d3a9c27d09d3b40dc65d01a774825a
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |82 |82 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |650 |650 |0 |
|test (roxie) |768 |768 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   33 sec (00:00:33) |  182 sec (00:03:02) |   55 sec (00:00:55) |    6 sec (00:00:06) |   36 sec (00:00:36) | 1097 sec (00:18:17) |   31 sec (00:00:31) | 1440 sec (00:24:00) |

richardkchapman commented:
@ghalliday Please review/merge

richardkchapman commented:
@dcamper Your latest commit seems to have been squashed so that the title only refers to the last change you made, rather than the overall impact. In order to get the changelog looking nice, please update the commit title to

HPCC-16883 Add time zone support to Std.Date 

or similar

richardkchapman commented:
@dcamper Needs rebasing for a clean merge (and commit message needs fixing)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: aed566b645f3c030ec84c278fd4cf759b87ed396
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |706 |706 |0 |
|test (roxie) |819 |819 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   20 sec (00:00:20) |  182 sec (00:03:02) |   55 sec (00:00:55) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1170 sec (00:19:30) |   17 sec (00:00:17) | 1472 sec (00:24:32) |

richardkchapman commented:
@ghalliday Any concerns with this?

ghalliday commented on ecllibrary/std/Date.ecl:
```diff
@@ -1354,4 +1354,777 @@ EXPORT DateTime_rec CreateDateTimeFromSeconds(Seconds_t seconds) := TRANSFORM
+ *                      unique location names.
+ */
+
+EXPORT UniqueTZLocations() := FUNCTION
```
This should have a test case.

ghalliday commented on ecllibrary/teststd/Date/TestDate.ecl:
```diff
@@ -237,6 +241,18 @@ EXPORT TestDate := MODULE
     ASSERT(Date.SecondsFromDateTimeRec(vCreateDateTime) = 917872496);
 
+    ASSERT(EXISTS(Date.TimeZone.UniqueTZAbbreviations()));
+    ASSERT(EXISTS(Date.TimeZone.TZDataForLocation('NORTH AMERICA')));
```
Better to have a slightly better test - e.g., count the number of military time zones = 25
(I never knew about half these timezones before this PR!)

ghalliday commented on ecllibrary/teststd/Date/TestDate.ecl:
```diff
@@ -237,6 +241,18 @@ EXPORT TestDate := MODULE
+
+    ASSERT(Date.TimeZone.AppendTZOffset(vTZInputSample, tz, seconds_to_utc)[1].seconds_to_utc_is_valid = TRUE);
+    ASSERT(Date.TimeZone.AppendTZOffset(vTZInputSample, tz, seconds_to_pt, toTimeZoneAbbrev := 'PT', toLocation := 'NORTH AMERICA')[1].seconds_to_pt_is_valid = TRUE);
+    ASSERT(Date.TimeZone.AppendTZAdjustedTime(vTZInputSample, time, tz, utc_time)[1].utc_time_is_valid = TRUE);
```
It would also be better to check more of the results

ghalliday reviewed (commented):
This looks like a really good PR.  Main comments are that it could do with a few more test cases - check all functions are tested, some more edge cases checked, and slightly more comprehensive tests.

ghalliday commented:
@dcamper see comment

dcamper commented on ecllibrary/std/Date.ecl:
```diff
@@ -1354,4 +1354,777 @@ EXPORT DateTime_rec CreateDateTimeFromSeconds(Seconds_t seconds) := TRANSFORM
+    // Flatten collected names, so there is one name per record
+    flattenedNames := NORMALIZE
+        (
+            collectedNames(EXISTS(names)),
```
Good catch.  That filter was left over from a version where not all time zones had location names.

dcamper commented:
@ghalliday All comments addressed; please re-review.

Note:  There was one data change in this latest commit.  While creating tests I discovered that the TZ abbreviation 'IST' in location 'ASIA' was actually used twice for two entirely different time zones.  The original source of the tz information has that duplicate as well.  In order to differentiate those duplicate time zones, I changed the location of the two Israel-specific zones to 'ISRAEL'.  The location name is not part of a standard as far as I can tell, and this seemed to be the minimum change I could make to provide differentiation.  I have added comments regarding the data source and this deviation from that source to the file.  If you have any thoughts on this change, please let me know.

Thanks!

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 82e9d3aa852f29301bf911262e872203bfea4676
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   25 sec (00:00:25) |  185 sec (00:03:05) |   56 sec (00:00:56) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1154 sec (00:19:14) |   16 sec (00:00:16) | 1464 sec (00:24:24) |
